test/core/handshake/sniffing_handshaker_server_ssl.c

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ * Copyright 2016 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <arpa/inet.h>
+#include <openssl/err.h>
+#include <openssl/ssl.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include <grpc/grpc.h>
+#include <grpc/grpc_security.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/sync.h>
+#include <grpc/support/thd.h>
+#include "src/core/lib/iomgr/load_file.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+
+#include "src/core/lib/channel/handshaker_factory.h"
+#include "src/core/lib/channel/handshaker_registry.h"
+#include "src/core/lib/security/transport/security_handshaker.h"
+
+#include "test/core/handshake/server_ssl_common.h"
+
+static void fake_sniffing_handshaker_destroy(grpc_exec_ctx *ctx,
+                                             grpc_handshaker *handshaker);
+static void fake_sniffing_handshaker_shutdown(grpc_exec_ctx *ctx,
+                                              grpc_handshaker *handshaker,
+                                              grpc_error *error);
+static void fake_sniffing_handshaker_do_handshake(
+    grpc_exec_ctx *ctx, grpc_handshaker *handshaker,
+    grpc_tcp_server_acceptor *acceptor, grpc_closure *on_handshake_done,
+    grpc_handshaker_args *args);
+
+const grpc_handshaker_vtable fake_sniffing_handshaker_vtable = {
+    fake_sniffing_handshaker_destroy, fake_sniffing_handshaker_shutdown,
+    fake_sniffing_handshaker_do_handshake};
+
+typedef struct {
+  grpc_handshaker base;
+  gpr_mu mu;
+  grpc_closure on_read;
+  grpc_closure *on_handshake_done;
+  grpc_handshaker_args *args;
+  grpc_slice_buffer read_scratch_buffer;
```

Comment:
I don't think this field is necessary.  We can use `h->args->read_buffer` directly in the call to `grpc_endpoint_read()`.

As a consequence of this, `fake_sniffing_handshaker_on_read()` is no longer needed, since its only real purpose is to copy the slices from `h->read_scratch_buffer` into `h->args->read_buffer`.  Instead, we can use `on_handshake_done` as the callback we pass to `grpc_endpoint_read()`.

And once we eliminate `fake_sniffing_handshaker_on_read()`, then we can eliminate all of the data fields in this struct, because their only purpose was to retain state between `fake_sniffing_handshaker_do_handshake()` and `fake_sniffing_handshaker_on_read()`.  And if there are no data members, then there's no need for a mutex, which means that this entire struct is unnecessary.

In other words, I think we can just use `grpc_handshaker` directly instead of creating our own struct here.

Reply:
Good point! removed the struct